test/rule_test_support.dart

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/error/error.dart';
+import 'package:analyzer/file_system/file_system.dart';
+import 'package:analyzer/src/dart/analysis/analysis_context_collection.dart';
+import 'package:analyzer/src/dart/analysis/byte_store.dart';
+import 'package:analyzer/src/dart/analysis/driver_based_analysis_context.dart';
+import 'package:analyzer/src/dart/analysis/experiments.dart';
+import 'package:analyzer/src/test_utilities/mock_packages.dart';
+import 'package:analyzer/src/test_utilities/mock_sdk.dart';
+import 'package:analyzer/src/test_utilities/package_config_file_builder.dart';
+import 'package:analyzer/src/test_utilities/resource_provider_mixin.dart';
+import 'package:linter/src/rules.dart';
+import 'package:meta/meta.dart';
+import 'package:test/test.dart';
+
+export 'package:analyzer/src/test_utilities/package_config_file_builder.dart';
+
+class AnalysisOptionsFileConfig {
+  final List<String> experiments;
+  final List<String> lints;
+
+  AnalysisOptionsFileConfig({
+    this.experiments = const [],
+    this.lints = const [],
+  });
+
+  String toContent() {
+    var buffer = StringBuffer();
+
+    buffer.writeln('analyzer:');
+    buffer.writeln('  enable-experiment:');
+    for (var experiment in experiments) {
+      buffer.writeln('    - $experiment');
+    }
+
+    buffer.writeln('linter:');
+    buffer.writeln('  rules:');
+    for (var lint in lints) {
+      buffer.writeln('    - $lint');
+    }
+
+    return buffer.toString();
+  }
+}
+
+abstract class LintRuleTest extends PubPackageResolutionTest {
+  String? get lintRule;
+
+  @override
+  List<String> get _lintRules => [if (lintRule != null) lintRule!];
+
+  Future<void> assertHasLint(String code) async {
+    addTestFile(code);
+    await resolveTestFile();
+
+    for (var error in errors) {
```

Comment:
There's an interesting design question here. Currently this only validates that the lint rule produced at least one diagnostic. It doesn't verify that no other diagnostics (lint or otherwise) were produced. The tests would be stronger if we completely controlled the set of diagnostics.

Longer term, consider copying `GatheringErrorListener` from either `analyzer` or `analysis_server` (or moving the analyzer version into `src`). It does a really good job of comparing expected outcomes and producing a useful report when they don't match (including the code needed to update the test if the test expectations need to be changed.

It also allows you to specify the location of the diagnostic, which would be useful.

Reply:
Absolutely agreed. I think we'll want to do that for sure when we get to migrating integration tests.